### PR TITLE
Added function to init library

### PR DESCRIPTION
**Pull request by luisggpina**

### CODE REVIEW

File: src/CodeGen.hs [hunk 1]
```diff
@@ -173,14 +175,25 @@ endVal mnem args c
...
+
 initFps :: [Mapping] -> Doc
 initFps mappings
-  = text "state_func_t (*start_state[329])(struct syscall_ *) = " <>
-    encloseSep lbrace rbrace comma (map initFp syscallInfo) <> text ";\n"
-  where initFp (n, i)
-          = case findIndex (\(((Name name), _, _, _),_) -> n == name) mappings of
-              Just _  -> text ("&init_" ++ n)
-              Nothing -> text "&init_default_match"
+  = let (_,res) = (foldl' initFp (0,text "") syscallInfo) in res
```
Comment (daniel-grumberg) [inline src/CodeGen.hs]: Why foldl' here? We are only ever gonna fold the size of syscallInfo which won't exceed 300-something elements, the sequential nature of foldl' is not needed here AFAIK

Comment (luisggpina) [inline src/CodeGen.hs]: The idea is to generate lines like:
```
   start_state[0] = &init_default_match;
   start_state[1] = &init_write;
   start_state[2] = &init_default_match;
   start_state[3] = &init_default_match;
```

For that, I need to know the index of each function.  I decided to fold over a tuple `(Integer,Doc)` and, at each step, I increment the `Integer` and append stuff to the `Doc`.

Is there a more elegant way of implementing this?  I was thinking of using a state monad or a reader monad, but my monadness is weak..

Comment (daniel-grumberg) [inline src/CodeGen.hs]: No, I what you are doing is great (No need to use a State Monad for a single integer). I don't know if you made a typo, but my point is why are you using `foldl'` over `foldl`. The difference is that `foldl` is lazy as opposed to `foldl'` being strict. My point was that using vanilla `foldl` is more idiomatic as opposed to forcing the evaluation to be strict with `foldl'`. For more info about the difference check out [https://wiki.haskell.org/Foldr_Foldl_Foldl'](url)

Comment (luisggpina) [inline src/CodeGen.hs]: Oh, I see.  That makes sense, I'll replace `fold'` with `fold`.

I read somewhere online that `fold'` performed much better than `fold` and I deduced it should be preferred.  Sorry for my confusion and thanks for the help.

File: src/CodeGen.hs [hunk 2]
```diff
@@ -173,14 +175,25 @@ endVal mnem args c
...
+
 initFps :: [Mapping] -> Doc
 initFps mappings
-  = text "state_func_t (*start_state[329])(struct syscall_ *) = " <>
-    encloseSep lbrace rbrace comma (map initFp syscallInfo) <> text ";\n"
-  where initFp (n, i)
-          = case findIndex (\(((Name name), _, _, _),_) -> n == name) mappings of
-              Just _  -> text ("&init_" ++ n)
-              Nothing -> text "&init_default_match"
+  = let (_,res) = (foldl initFp (0,text "") syscallInfo) in res
```
Comment (luisggpina) [inline src/CodeGen.hs]: Fixed (through history rewrite to keep the history clean after merging)

Merge at your discretion

File: src/CodeGen.hs (hunk 1 above)
Comment (daniel-grumberg) [inline src/CodeGen.hs]: You mean `foldl` ;), Once you do that I'll merge it in and update the tests accordingly

Comment (luisggpina) [inline src/CodeGen.hs]: I've updated to `foldl`, the code we are commenting on is outdated.  Check the latest version.  I've rewritten the history to look like `foldl` was there in the first place, to clean the history.

**Comment by daniel-grumberg** [comment]
Fixed the tests, did not think it warranted a PR (literally adding a line to call the new `init_div` at each test) so just went ahead and pushed to master

**Pull request merged**